Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp

```diff
@@ -935,6 +944,106 @@ class ObjectAllocationSinkingPhase : public Phase {
+            //    into a phantom node.
+            //
+            // 6. **If any other node in the program uses D@1 as an input (other than the cases listed above),**
+            //    **sinking will be disabled.** This ensures that no other operation:
+            //    - Adds new properties to the array.
+            //    - Modifies the array’s length.
+            //    - Updates any existing fields in a way that would break our assumptions about the array.
+            //    If such an operation exists, we **must not** sink the allocation, as it would lead to
+            //    incorrect behavior.
+            if (Options::useArrayAllocationSinking()) {
+                if (!m_graph.isWatchingArrayPrototypeChainIsSaneWatchpoint(node))
+                    goto escapeChildren;
+                target = &m_heap.newAllocation(node, Allocation::Kind::Array);
+
+                if (!m_sinkCandidates.contains(node) || !enableLengthMaterialization)
+                    goto escapeChildren;
+                Node* length = m_insertionSet.insertNode(m_nodeIndex, SpecInt32Only, JSConstant, node->origin,
+                    OpInfo(m_graph.freeze(jsNumber(node->newArraySize()))));
+                writes.add(PromotedLocationDescriptor(ArrayLengthPropertyPLoc), LazyNode(length));
+            } else
+                goto escapeChildren;
+            break;
+
+        case GetButterfly:
+        case GetArrayLength: {
+            Node* base = node->child1().node();
+            target = m_heap.onlyLocalAllocation(base);
+            if (target && target->isArrayAllocation()) {
+                if (node->op() == GetArrayLength && m_sinkCandidates.contains(base))
+                    exactRead = PromotedLocationDescriptor(ArrayLengthPropertyPLoc);
+            } else
+                goto escapeChildren;
+            break;
+        }
+
+        case GetByVal:
+        case PutByVal: {
+            ArrayMode arrayMode = node->arrayMode();
+            Node* base = m_graph.varArgChild(node, 0).node();
+            Node* index = m_graph.varArgChild(node, 1).node();
+            target = m_heap.onlyLocalAllocation(base);
+
+            if (target && target->isArrayAllocation()
+                && arrayMode.isInBounds()
+                && isNewArrayWithConstantSizeIndexingType(base->indexingType())
+                && index->isInt32Constant()) {
+                ASSERT_UNUSED(arrayMode, arrayMode.type() == Array::Int32 || arrayMode.type() == Array::Double || arrayMode.type() == Array::Contiguous);
+                if (node->op() == PutByVal) {
+                    Edge value = m_graph.varArgChild(node, 2);
+                    writes.add(PromotedLocationDescriptor(ArrayIndexedPropertyPLoc, index->asInt32()), LazyNode(value.node()));
+                    if (m_sinkCandidates.contains(base))
+                        m_nodeToUseKind.add(value.node(), value.useKind());
```

Comment:
I think OSR exit via this check in materialization code sounds not correct. Rather, we should emit Check with this UseKind here (so, I don't think we need  m_nodeToUseKind)

Reply:
The reason to update UseKing for edges of `MaterializeNewArrayWithConstantSize` is not only for the speculate check in the compile phase. But also for passing the validation checks for double edges of `MaterializeNewArrayWithConstantSize` when the indexing type is in the double cases.

Reply:
Updated with an inserted `Check`.